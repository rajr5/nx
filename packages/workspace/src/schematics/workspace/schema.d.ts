import { UnitTestRunner } from '../../utils/test-runners';

export interface Schema {
  directory: string;
  name: string;
  npmScope?: string;
  skipInstall?: boolean;
  skipGit?: boolean;
  style?: string;
  commit?: { name: string; email: string; message?: string };
  cli: 'nx' | 'angular';
}
