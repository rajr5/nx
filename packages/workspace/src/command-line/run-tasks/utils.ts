import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';
import { NxJson, readNxJson, readWorkspaceJson } from '../shared';
import { WorkspaceResults } from '../workspace-results';
import { ProjectGraphNode } from '../project-graph';

export interface Arguments<T extends yargs.Arguments> {
  nxArgs: T;
  targetArgs: T;
  overrides: any;
}

const ignoreArgs = ['$0', '_'];
export function splitArgs<T extends yargs.Arguments>(
  args: T,
  nxSpecific: (keyof T)[]
): Arguments<T> {
  const nx: any = {};
  const targetArgs: any = {};

  const overrides = yargsParser(args._.slice(1));
  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    if (nxSpecific.includes(key as any)) {
      nx[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      targetArgs[key] = value;
    }
  });

  return { nxArgs: nx, targetArgs, overrides };
}
export interface TaskArgs {
  projects: string[];
  target: string;
  all?: boolean;
}

export function projectHasTargetAndConfiguration(
  project: ProjectGraphNode,
  target: string,
  configuration?: string
) {
  if (
    !project.data ||
    !project.data.architect ||
    !project.data.architect[target]
  ) {
    return false;
  }

  if (!configuration) {
    return !!project.data.architect[target];
  } else {
    return (
      project.data.architect[target].configurations &&
      project.data.architect[target].configurations[configuration]
    );
  }
}

export interface Environment {
  nxJson: NxJson;
  workspaceJson: any;
  workspace: any;
}

export function readEnvironment(target: string): Environment {
  const nxJson = readNxJson();
  const workspaceJson = readWorkspaceJson();
  const workspace = new WorkspaceResults(target);

  return { nxJson, workspaceJson, workspace };
}
