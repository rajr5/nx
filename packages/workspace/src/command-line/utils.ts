import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';

/**
 * These options are only for getting an array with properties of AffectedOptions.
 *
 * @remark They are not defaults or useful for anything else
 */
const dummyOptions: NxArgs = {
  target: '',
  configuration: '',
  runner: '',
  parallel: false,
  maxParallel: 0,
  'max-parallel': 0,
  untracked: false,
  uncommitted: false,
  all: false,
  base: 'base',
  head: 'head',
  exclude: ['exclude'],
  files: [''],
  onlyFailed: false,
  'only-failed': false,
  verbose: false,
  help: false,
  version: false,
  quiet: false,
  plain: false,
  withDeps: false,
  'with-deps': false,
  projects: [],
  select: ''
} as any;

const nxSpecific = Object.keys(dummyOptions);

export interface NxArgs {
  target?: string;
  configuration?: string;
  runner?: string;
  parallel?: boolean;
  maxParallel?: number;
  'max-parallel'?: number;
  untracked?: boolean;
  uncommitted?: boolean;
  all?: boolean;
  base?: string;
  head?: string;
  exclude?: string[];
  files?: string[];
  onlyFailed?: boolean;
  'only-failed'?: boolean;
  verbose?: boolean;
  help?: boolean;
  version?: boolean;
  quiet?: boolean;
  plain?: boolean;
  withDeps?: boolean;
  'with-deps'?: boolean;
  projects?: string[];
  select?: string;
  _: string[];
}

const ignoreArgs = ['$0', '_'];

export function splitArgsIntoNxArgsAndOverrides(
  args: yargs.Arguments
): { nxArgs: NxArgs; overrides: yargs.Arguments } {
  const nxArgs: any = {};
  const overrides = yargsParser(args._);
  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    if (nxSpecific.includes(key as any)) {
      nxArgs[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      overrides[key] = value;
    }
  });

  if (!nxArgs.projects) {
    nxArgs.projects = [];
  } else {
    nxArgs.projects = args.projects.split(',').map((p: string) => p.trim());
  }

  return { nxArgs, overrides };
}
