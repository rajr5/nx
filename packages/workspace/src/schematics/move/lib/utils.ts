import { WorkspaceDefinition } from '@angular-devkit/core/src/workspace';
import * as path from 'path';
import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 *
 * @param schema
 * @param workspace
 */
export function getDestination(
  schema: Schema,
  workspace: WorkspaceDefinition | any
): string {
  const project = workspace.projects.get
    ? workspace.projects.get(schema.projectName)
    : workspace.projects[schema.projectName];
  const projectType = project.extensions
    ? project.extensions['projectType']
    : project.projectType;

  let rootFolder = 'libs';
  if (projectType === 'application') {
    rootFolder = 'apps';
  }
  return path.join(rootFolder, schema.destination).split(path.sep).join('/');
}

/**
 * Replaces slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  return path.replace(/\//g, '-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizeSlashes(input: string): string {
  return input
    .split('/')
    .filter((x) => !!x)
    .join('/');
}
