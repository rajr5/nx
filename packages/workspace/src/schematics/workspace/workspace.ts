import {
  apply,
  branchAndMerge,
  chain,
  mergeWith,
  Rule,
  SchematicContext,
  template,
  Tree,
  url
} from '@angular-devkit/schematics';
import { Schema } from './schema';
import { strings } from '@angular-devkit/core';
import {
  angularCliVersion,
  prettierVersion,
  typescriptVersion,
  nxVersion
} from '../../utils/versions';

export const DEFAULT_NRWL_PRETTIER_CONFIG = {
  singleQuote: true
};

export default function(options: Schema): Rule {
  if (!options.name) {
    throw new Error(`Invalid options, "name" is required.`);
  }

  return (host: Tree, context: SchematicContext) => {
    const npmScope = options.npmScope ? options.npmScope : options.name;
    const templateSource = apply(url('./files'), [
      template({
        utils: strings,
        dot: '.',
        tmpl: '',
        typescriptVersion,
        prettierVersion,
        angularCliVersion,
        nxVersion,
        ...(options as object),
        npmScope,
        defaultNrwlPrettierConfig: JSON.stringify(
          DEFAULT_NRWL_PRETTIER_CONFIG,
          null,
          2
        )
      })
    ]);
    return chain([branchAndMerge(chain([mergeWith(templateSource)]))])(
      host,
      context
    );
  };
}
