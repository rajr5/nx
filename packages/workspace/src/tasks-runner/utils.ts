import { Task } from './tasks-runner';
import { ProjectGraphNode } from '../core/project-graph';

const commonCommands = ['build', 'test', 'lint', 'e2e', 'deploy'];

export function getCommand(cliCommand: string, isYarn: boolean, task: Task) {
  const args = Object.entries(task.overrides || {})
    .map(([prop, value]) => `--${prop}=${value}`)
    .join(' ');

  if (commonCommands.includes(task.target.target)) {
    const config = task.target.configuration
      ? `--configuration ${task.target.configuration} `
      : '';
    return `${cliCommand}${isYarn ? '' : ' --'} ${task.target.target} ${
      task.target.project
    } ${config} ${args}`.trim();
  } else {
    const config = task.target.configuration
      ? `:${task.target.configuration} `
      : '';
    return `${cliCommand}${isYarn ? '' : ' --'}  run ${task.target.project}:${
      task.target.target
    }${config} ${args}`.trim();
  }
}

export function getOutputs(p: Record<string, ProjectGraphNode>, task: Task) {
  const architect = p[task.target.project].data.architect[task.target.target];
  let opts = architect.options || {};
  if (
    architect.configurations &&
    architect.configurations[task.target.configuration]
  ) {
    opts = {
      ...opts,
      ...architect.configurations[task.target.configuration]
    };
  }
  let outputs = [];
  if (opts.outputPath) {
    outputs.push(opts.outputPath);
  }
  return outputs;
}
