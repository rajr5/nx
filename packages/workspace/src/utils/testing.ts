import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../collection.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationTestRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
