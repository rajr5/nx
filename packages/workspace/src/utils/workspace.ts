import { Tree, Rule } from '@angular-devkit/schematics';
import { workspaces } from '@angular-devkit/core';

function createHost(tree: Tree): workspaces.WorkspaceHost {
  return {
    async readFile(path: string): Promise<string> {
      const data = tree.read(path);
      if (!data) {
        throw new Error('File not found.');
      }

      return data.toString();
    },
    async writeFile(path: string, data: string): Promise<void> {
      return tree.overwrite(path, data);
    },
    async isDirectory(path: string): Promise<boolean> {
      // approximate a directory check
      return !tree.exists(path) && tree.getDir(path).subfiles.length > 0;
    },
    async isFile(path: string): Promise<boolean> {
      return tree.exists(path);
    },
  };
}
export async function getWorkspace(tree: Tree, path = '/') {
  const host = createHost(tree);

  const { workspace } = await workspaces.readWorkspace(path, host);

  return workspace;
}
export function updateWorkspace(
  updater: (
    workspace: workspaces.WorkspaceDefinition
  ) => void | PromiseLike<void>
): Rule;
export function updateWorkspace(
  workspace: workspaces.WorkspaceDefinition
): Rule;
export function updateWorkspace(
  updaterOrWorkspace:
    | workspaces.WorkspaceDefinition
    | ((workspace: workspaces.WorkspaceDefinition) => void | PromiseLike<void>)
): Rule {
  return async (tree: Tree) => {
    const host = createHost(tree);

    if (typeof updaterOrWorkspace === 'function') {
      const { workspace } = await workspaces.readWorkspace('/', host);

      const result = updaterOrWorkspace(workspace);
      if (result !== undefined) {
        await result;
      }

      await workspaces.writeWorkspace(workspace, host);
    } else {
      await workspaces.writeWorkspace(updaterOrWorkspace, host);
    }
  };
}
