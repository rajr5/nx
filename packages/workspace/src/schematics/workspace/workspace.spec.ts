import { Tree } from '@angular-devkit/schematics';
import { NxJson, readJsonInTree } from '@nrwl/workspace';
import { runSchematic } from '../../utils/testing';

describe('workspace', () => {
  let projectTree: Tree;

  beforeEach(() => {
    projectTree = Tree.empty();
  });

  it('should create files', async () => {
    const tree = await runSchematic('workspace', { name: 'proj' }, projectTree);
    expect(tree.exists('/nx.json')).toBe(true);
    expect(tree.exists('/workspace.json')).toBe(true);
    expect(tree.exists('/.prettierrc')).toBe(true);
    expect(tree.exists('/.prettierignore')).toBe(true);
  });

  it('should create nx.json', async () => {
    const tree = await runSchematic('workspace', { name: 'proj' }, projectTree);
    const nxJson = readJsonInTree<NxJson>(tree, '/nx.json');
    expect(nxJson).toEqual({
      npmScope: 'proj',
      implicitDependencies: {
        'workspace.json': '*',
        'package.json': {
          dependencies: '*',
          devDependencies: '*'
        },
        'tsconfig.json': '*',
        'tslint.json': '*',
        'nx.json': '*'
      },
      projects: {}
    });
  });

  it('should recommend vscode extensions', async () => {
    const tree = await runSchematic('workspace', { name: 'proj' }, projectTree);
    const recommendations = readJsonInTree<{ recommendations: string[] }>(
      tree,
      '/.vscode/extensions.json'
    ).recommendations;

    expect(recommendations).toEqual([
      'ms-vscode.vscode-typescript-tslint-plugin',
      'esbenp.prettier-vscode'
    ]);
  });

  it('should recommend vscode extensions (angular)', async () => {
    const tree = await runSchematic(
      'workspace',
      { name: 'proj', cli: 'angular' },
      projectTree
    );
    const recommendations = readJsonInTree<{ recommendations: string[] }>(
      tree,
      '/.vscode/extensions.json'
    ).recommendations;

    expect(recommendations).toEqual([
      'nrwl.angular-console',
      'angular.ng-template',
      'ms-vscode.vscode-typescript-tslint-plugin',
      'esbenp.prettier-vscode'
    ]);
  });
});
