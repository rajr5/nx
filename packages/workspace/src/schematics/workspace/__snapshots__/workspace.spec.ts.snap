// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workspace should create a prettierrc file 1`] = `
"{
  \\"singleQuote\\": true
}"
`;
